Extension/src/Utility/Eventing/eventParser.ts

```diff
@@ -0,0 +1,265 @@
+    return [isSync, once, filters, source];
+
+}
+
+const sandbox = new Sandbox();
+
+function generateFilterFn(scanner: Scanner): Filter {
+    // take all the tokens until we hit a closing bracket
+    const inner = [...scanner.takeUntil(Kind.CloseBracket, {
+        nestable: [[Kind.OpenBracket, Kind.CloseBracket]],
+        escape: [Kind.Backslash]
+    })];
+
+    // a filter expression is one or more of:
+    //  - a regular expression
+    //  - a javascript expression
+    // separated by && or ||
+    const expression = new Array<string>();
+
+    // remove leading whitespace
+    eatWhitespace(inner);
+
+    outer:
+    while (inner.length) {
+        eatWhitespace(inner);
+        let token = inner.shift()!;
+
+        switch (token.kind) {
+            case Kind.Slash: {
+                // regular expression
+                // take all tokens until we hit a slash that isn't escaped
+                const rxExpression = [token];
+
+                while (inner.length) {
+
+                    token = inner.shift()!;
+                    rxExpression.push(token); // store this token as part of the regex
+
+                    switch (token.kind) {
+                        case Kind.Slash:
+                            // now see if there are any flags
+                            if (inner.length) {
+                                token = inner[0];
+                                if (token.kind === Kind.Identifier) {
+                                    rxExpression.push(inner.shift()!);
+                                }
+                            }
+                            // at this point, we should have the whole regex.
+                            // turn this into an expression that tests against any $strings
+                            expression.push(`(!!$strings.find( ($text)=> { const r = ${rxExpression.map(token => token.text).join('')}.exec($text); if( r ) { $captures.push( ...r ); return true; } } ))`);
+                            continue outer;
+
+                        case Kind.Backslash:
+                            // take the next token, and add it to the regex
+                            rxExpression.push(inner.shift()!);
+                            continue;
+                    }
+
+                    // if we have nothing left, and we haven't closed the regex, then it's an error
+                    if (!inner.length) {
+                        throw new Error(`unterminated regular expression ${inner}`);
+                    }
+                }
+            }
+                break;
+
+            case Kind.OpenParen:
+            case Kind.CloseParen:
+            case Kind.BarBar:
+            case Kind.AmpersandAmpersand:
+                // add that token to the final expression
+                expression.push(token.text!);
+                break;
+
+            case Kind.StringLiteral:
+                // if we're given a string literal and the next token after that is one of {||, &&, ],}

```

Comment:
Is this correct? It appears to be a `{ }` containing a comma separated list of `||`, `&&`, `]` and then an extra `,`?

Reply:
removed the extra comma